### PR TITLE
Use response executor when handling failure in field-caps

### PR DESCRIPTION
**Pull request by dnhatn**
The transport service may call onFailure with a thread pool other than search_coordinator. This change adds a workaround to ensure that the merging of field-caps always occurs on the search_coordinator.

### CODE REVIEW

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
I think we already dispatch to the right executor on a regular response, so this change will make us fork twice. But I can see that we might not dispatch to the right executor on a connection error. That seems like something we should fix in the remote-cluster client rather than here.

**Comment by dnhatn** [comment]
> I think we already dispatch to the right executor on a regular response, so this change will make us fork twice. But I can see that we might not dispatch to the right executor on a connection error.

@DaveCTurner Thanks for looking. We use the generic threadpool for notifying remote connection errors. I pushed https://github.com/elastic/elasticsearch/pull/107370/commits/2e4c59ac1e0b4a0daede6e80a75a5e18ddc60283.

**Review by DaveCTurner — CHANGES_REQUESTED** [review]
2e4c59ac1e0b4a0daede6e80a75a5e18ddc60283 is still going to double-fork on most exceptions (see e.g. `org.elasticsearch.transport.TransportService#getInternalSendExceptionExecutor` and `org.elasticsearch.transport.InboundHandler#handleException`). Let's just fork the cases that don't already fork.

**Review by DaveCTurner — COMMENTED** [review]
Oh wait sorry I misread it, I see we're only forking exceptions from `maybeEnsureConnected`. Hmm this still isn't quite right tho, there's also `org.elasticsearch.transport.TransportService#handleSendRequestException`.

**Review by DaveCTurner — COMMENTED** [review]
Hmm I'm not too happy about introducing an `ActionListener#wrap` here (double-completion bugs abound) nor really about replicating most of that exception-dispatching `ActionRunnable` in three places. I think I'd like to spend a little more time on coming up with a better solution here but unfortunately I don't have that time right now.

Could you keep this open as a placeholder for that work, and/or open a bug report issue for me to follow up on, and go back to the original workaround of just handling this in the (exception path of) the field caps action?

**Comment by dnhatn** [comment]
Thanks, David. I've narrowed down this PR to just field-caps. This should unblock our work in ES|QL. I'll also open an issue for the transport service. Could you give it another look?

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java [hunk 1]
```diff
@@ -252,7 +253,13 @@ private void doExecuteForked(Task task, FieldCapabilitiesRequest request, final
                 remoteClusterClient.execute(
                     TransportFieldCapabilitiesAction.REMOTE_TYPE,
                     remoteRequest,
-                    ActionListener.releaseAfter(remoteListener, refs.acquire())
+                    // The underlying transport service may call onFailure with a thread pool other than search_coordinator.
+                    // This fork is a workaround to ensure that the merging of field-caps always occurs on the search_coordinator.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:257]: nit: could you link from this comment to the issue that tracks the bug around which we're working?

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit)

**Pull request merged**